go/test/endtoend/vtgate/queries/random/query_gen_test.go

```diff
@@ -0,0 +1,511 @@
+	}
+
+	// add them to newTable
+	newTable.addColumns(grouping...)
+	newTable.addColumns(aggregates...)
+
+	// add new table to schemaTables
+	newTable.name = sqlparser.NewDerivedTable(false, sel)
+	schemaTables = append(schemaTables, newTable)
+
+	// derived tables (partially unsupported)
+	// TODO: derived tables fails a lot
+	if rand.Intn(10) < 1 && TestFailingQueries {
+		sel = randomQuery(schemaTables, 3, 3)
+	}
+
+	return sel
+}
+
+func createTablesAndJoin(schemaTables []tableT, sel *sqlparser.Select) ([]tableT, bool) {
+	var tables []tableT
+	// add at least one of original emp/dept tables for now because derived tables have nil columns
+	tables = append(tables, schemaTables[rand.Intn(2)])
+
+	sel.From = append(sel.From, newAliasedTable(tables[0], "tbl0"))
+	tables[0].setName("tbl0")
+
+	numTables := rand.Intn(len(schemaTables))
+	for i := 0; i < numTables; i++ {
+		tables = append(tables, randomEl(schemaTables))
+		sel.From = append(sel.From, newAliasedTable(tables[i+1], fmt.Sprintf("tbl%d", i+1)))
+		tables[i+1].setName(fmt.Sprintf("tbl%d", i+1))
+	}
+
+	// TODO: outer joins produce mismatched results
+	isJoin := rand.Intn(2) < 1 && TestFailingQueries
+	if isJoin {
+		newTable := randomEl(schemaTables)
+		tables = append(tables, newTable)
+
+		// create the join before aliasing
+		newJoinTableExpr := createJoin(tables, sel)
+
+		// alias
+		tables[numTables+1].setName(fmt.Sprintf("tbl%d", numTables+1))
+
+		// create the condition after aliasing
+		newJoinTableExpr.Condition = sqlparser.NewJoinCondition(sqlparser.AndExpressions(createWherePredicates(tables, true)...), nil)
+		sel.From[numTables] = newJoinTableExpr
+	}
+
+	return tables, isJoin
+}
+
+// creates a left join (without the condition) between the last table in sel and newTable
+// tables should have one more table than sel
+func createJoin(tables []tableT, sel *sqlparser.Select) *sqlparser.JoinTableExpr {
+	n := len(sel.From)
+	if len(tables) != n+1 {
+		log.Fatalf("sel has %d tables and tables has %d tables", len(sel.From), n)
+	}
+
+	return sqlparser.NewJoinTableExpr(sel.From[n-1], sqlparser.LeftJoinType, newAliasedTable(tables[n], fmt.Sprintf("tbl%d", n)), nil)
+}
+
+// returns the grouping columns as three types: sqlparser.GroupBy, sqlparser.SelectExprs, []column
+func createGroupBy(tables []tableT, maxGB int) (groupBy sqlparser.GroupBy, groupSelectExprs sqlparser.SelectExprs, grouping []column) {
+	numGBs := rand.Intn(maxGB)
+	for i := 0; i < numGBs; i++ {
+		tblIdx := rand.Intn(len(tables))
+		col := randomEl(tables[tblIdx].cols)
+		// TODO: grouping by a date column sometimes errors
+		if col.typ == "date" && !TestFailingQueries {
+			continue
+		}
+		groupBy = append(groupBy, newColumn(col))
+
+		// add to select
+		if rand.Intn(2) < 1 {
+			groupSelectExprs = append(groupSelectExprs, newAliasedColumn(col, fmt.Sprintf("cgroup%d", i)))
+			col.name = fmt.Sprintf("cgroup%d", i)
+			grouping = append(grouping, col)
+		}
+	}
+
+	return groupBy, groupSelectExprs, grouping
```

Comment:
You can omit the variable names as you're already naming your return variables in the function declaration.
```suggestion
	return
```